doc/arch/001-screen-controllers.md

```diff
@@ -0,0 +1,102 @@
+# ADR: Screen Controllers
+
+**Status**
+Accepted
+
+**Context**
+We don’t want to put business logic inside Android framework classes (like an `Activity` or `Fragment`) because those cannot be unit tested. To enable a fast feedback loop (i.e. tests that run on the JVM and not Android VM), we separate screens and controllers using the [MVI architecture](https://medium.com/@ragunathjawahar/mvi-series-a-pragmatic-reactive-architecture-for-android-87ca176387d1) [pattern](https://medium.com/@ragunathjawahar/mvi-series-a-pragmatic-reactive-architecture-for-android-87ca176387d1).
+
+**Decision**
+Every screen has one controller that consumes user events, performs business logic with the help of data repositories and communicates UI changes back to the screen.
+
+User interactions are abstracted inside data classes of type `UiEvent`. These events flow to the controller in the form of RxJava streams.
+
+```kotlin
+// Create the UsernameTextChanged event by listening to the EditText
+RxTextView
+  .textChanges(usernameEditText)
+  .map { text -> UsernameTextChanged(text) }
+
+// Event
+data class UsernameTextChanged(text: String) : UiEvent
+```
+
+The screen sends a single stream of UI events to the controller and gets back a transformed stream of UI changes. The flow of data is unidirectional. To merge multiple streams into one, RxJava’s `merge()`  operator is used.
+
```

Comment:
To understand this we expect that the reader knows what MVI and unidirectional data flow patterns.
Is it right?

Reply:
@pratul should we explain how we're achieving unidirectional data flow? Or is that upto the reader to understand from MVI?

Reply:
Hm, IMHO "unidirectional flow of data" is vaguely understandable, and if the reader wants to know more they could search around. For example, I found this: https://www.exclamationlabs.com/blog/the-case-for-unidirectional-data-flow/

I say this because we're also heavily using Rx and streams and we don't explain any of that either.